### PR TITLE
(#204) duplicate applicant

### PR DESCRIPTION
**Pull request by rathorevaibhav**
#204 

Changes include:

1. Instead of associating applicant with a job, there'll be multiple applications for an applicant that will be associated with a job.
2. Migration for `hr_applications` table (which is a many-to-many relationship table between applicant and job).
3. Seeder to make applications for every unique applicant and then deleting the duplicate applicants.

Changes related to this issue targeted in next (tomorrow's) PR:
1. A new applicant won't be created if he/she is applying again (for the same or different job). Instead, a new application will be created.
2. Reason for eligibility, resume, and some other fields can be different for each application. Hence these details will be stored in `hr_applications` table.
3. `hr_applicant_round` will be renamed to `hr_application_round`. Will reference application_id instead of applicant_id.
4. `hr_applicant_reviews` will be renamed to `hr_application_review`. Will reference application_round_id instead of applicant_round_id.
5. The applicant edit screen will be replaced by application edit screen.

This means right now, if an applicant applies again, a new applicant will still be added along with a new application.

Commands to run:
```
$ php artisan migrate
$ php artisan db:seed --class=MigrateApplicantJobReferenceSeeder
```

### CODE REVIEW

File: database/seeds/MigrateApplicantJobReferenceSeeder.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        foreach ($applicants as $applicant) {
+
+            // checking if the current applicant entry is duplicate
+            if (array_key_exists($applicant->email, $uniqueApplicants)) {
+
+                // find the original applicant with the same email. $applicant is duplicate of $originalApplicant
+                $originalApplicant = $uniqueApplicants[$applicant->email];
+
+                // add the job of this duplicate applicant as the job of original applicant
+                Application::create([
```
[the remark this answers is not in the document]
Comment (rathorevaibhav) [reply database/seeds/MigrateApplicantJobReferenceSeeder.php]: It's a seeder and I don't think we'll be using it again. Moreover, this is specific to our use case and won't be required for any further project installations. Probably we can delete this after it is used? I don't see any value in making further changes or optimizations in it. Thoughts?

File: database/seeds/MigrateApplicantJobReferenceSeeder.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+class MigrateApplicantJobReferenceSeeder extends Seeder
+{
+    /**
+     * Run the database seeds.
+     *
+     * @return void
+     */
+    public function run()
+    {
+        $applicants = Applicant::all();
```
[the remark this answers is not in the document]
Comment (rathorevaibhav) [reply database/seeds/MigrateApplicantJobReferenceSeeder.php]: Same thoughts for the seeder below

**Comment by rathorevaibhav** [comment]
@bohraji I had considered duplicity but it's looking that the next PR will be even bigger and will contain both structural and code level changes. I think it'll be better for me (and the team) to ship this in pieces. Waiting for feature completion will increase the deployment, review and test effort.

Moreover, I've talked this with @PrateekNarang about the duplicate entries between deployments. We agreed on cleaning them manually until the system supports doesn't handle them automatically. Considering that, the completion is my top priority.

File: app/Models/HR/ApplicantRound.php [hunk 3]
```diff
@@ -41,7 +41,7 @@ public function _update($attr, $type = 'new', $reviews = [], $nextRound = 0)
         }
 
         if ($nextRound) {
-            $scheduled_person = User::findByEmail($applicant->job->posted_by);
+            $scheduled_person = User::findByEmail($applicant->applications->first()->job->posted_by);
```
[the remark this answers is not in the document]
Comment (rathorevaibhav) [reply app/Models/HR/ApplicantRound.php]: Yes, while I was going through eloquent, I saw we can utilize distant relations using hasManyThrough. So far, IMO, hasMany (the syntax) is used in one-to-many relationships and in this case we have many-to-many (belongsToMany). I had a brief look though. I think I'll have to revisit the documentation and have more clarity. If we agree on shipping this PR, I think I'll target this in the next one.

File: database/seeds/MigrateApplicantJobReferenceSeeder.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+                // add the job of this duplicate applicant as the job of original applicant
+                Application::create([
+                    'hr_job_id' => $applicant->hr_job_id,
+                    'hr_applicant_id' => $originalApplicant->id,
+                    'created_at' => $applicant->created_at,
+                    'updated_at' => $applicant->updated_at,
+                ]);
+
+                foreach($applicant->applicantRounds as $applicantRound) {
```
[the remark this answers is not in the document]
Comment (rathorevaibhav) [reply database/seeds/MigrateApplicantJobReferenceSeeder.php]: Oh, I wasn't aware of that, directly deleting the relationships. Thanks. Let's discuss the PR situation and I'll see if we should do it here.

I'll take care of this moving forward in other areas though.

File: app/Models/HR/ApplicantRound.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (bohraji) [reply app/Models/HR/ApplicantRound.php]: Alright. We can do this in the next PR.

File: database/seeds/MigrateApplicantJobReferenceSeeder.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (bohraji) [reply database/seeds/MigrateApplicantJobReferenceSeeder.php]: Alright.

**Comment by rathorevaibhav** [comment]
This is not required anymore. All changes in this PR are being tracked in #210

**Pull request closed without merging**